# Project 2: Data Import - Working with Web APIs and JSON (Movies Dataset)

# Project Brief for Self-Coders

Here you´ll have the opportunity to code major parts of Project 2 on your own. If you need any help or inspiration, have a look at the Videos or the Jupyter Notebook with the full code. <br> <br>
Keep in mind that it´s all about __getting the right results/conclusions__. It´s not about finding the identical code. Things can be coded in many different ways. Even if you come to the same conclusions, it´s very unlikely that we have the very same code. 

## Importing Data from JSON files 

1. __Import__ the json files __blockbusters.json__, __blockbusters2.json__, __blockbusters3.json__ and load the datasets into Pandas DataFrames.


In [2]:
import pandas as pd
pd.options.display.max_columns = 30

In [3]:
df = pd.read_json('blockbusters.json', orient = 'records')
df.head()

title      id     revenue  \
0             Avengers: Endgame  299534  2797800564   
1                        Avatar   19995  2787965087   
2  Star Wars: The Force Awakens  140607  2068223624   
3        Avengers: Infinity War  299536  2046239637   
4                       Titanic     597  1845034188   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               belongs_to_collection  runtime  
0  {'id': 86311, 'name': 'The Avengers Collection...      181  
1  {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2  {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3  {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                               None      194

In [4]:
df2 = pd.read_json('blockbusters2.json' )
df2.head()

title      id     revenue  \
0             Avengers: Endgame  299534  2797800564   
1                        Avatar   19995  2787965087   
2  Star Wars: The Force Awakens  140607  2068223624   
3        Avengers: Infinity War  299536  2046239637   
4                       Titanic     597  1845034188   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               belongs_to_collection  runtime  
0  {'id': 86311, 'name': 'The Avengers Collection...      181  
1  {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2  {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3  {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                               None      194

In [5]:
df3 = pd.read_json('blockbusters3.json', orient = 'split')
df3.head()

title      id     revenue  \
0             Avengers: Endgame  299534  2797800564   
1                        Avatar   19995  2787965087   
2  Star Wars: The Force Awakens  140607  2068223624   
3        Avengers: Infinity War  299536  2046239637   
4                       Titanic     597  1845034188   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               belongs_to_collection  runtime  
0  {'id': 86311, 'name': 'The Avengers Collection...      181  
1  {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2  {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3  {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                               None      194

## Working with APIs and JSON (Part 1)

2. __Create an account__ on https://www.themoviedb.org/

3. Get your personal __API Key__

4. __API-Request__ (movie module): Load all available information for the movie with __movie id = 140607__ into a Pandas DataFrame. <br> See https://developers.themoviedb.org/3/movies/get-movie-details for more information

In [6]:
import config
import requests
api_key  = 'api_key=' + config.api_key
api_key

'api_key=f87c31e443414f56263b3306d1fada82'

In [7]:
movie_id = 140607

In [8]:
movie_api = "https://api.themoviedb.org/3/movie/{}?"
url = movie_api.format(movie_id) + api_key
url

'https://api.themoviedb.org/3/movie/140607?api_key=f87c31e443414f56263b3306d1fada82'

In [9]:
r = requests.get(url)
r

<Response [200]>

In [10]:
data = r.json()
type(data)

dict

In [11]:
data_df = pd.Series(data)
data_df = data_df.to_frame().T
data_df

adult                     backdrop_path  \
0  False  /k6EOrckWFuz7I4z4wiRwz8zsj4H.jpg   

                               belongs_to_collection     budget  \
0  {'id': 10, 'name': 'Star Wars Collection', 'po...  245000000   

                                              genres  \
0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                            homepage      id    imdb_id  \
0  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   

  original_language                original_title  \
0                en  Star Wars: The Force Awakens   

                                            overview popularity  \
0  Thirty years after defeating the Galactic Empi...     46.289   

                        poster_path  \
0  /wqnLdwVXoBjKibFRR5U3y0aDUhs.jpg   

                                production_companies  \
0  [{'id': 1, 'logo_path': '/o86DbpburjxrqAzEDhXZ...   

                                production_countries release_date     revenue  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   2015-12-15  2068223624   

  runtime                                   spoken_languages    status  \
0     136  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                         tagline                         title  video  \
0  Every generation has a story.  Star Wars: The Force Awakens  False   

  vote_average vote_count  
0          7.4      15717

In [24]:
pd.json_normalize(data, sep = "_")

adult                     backdrop_path     budget  \
0  False  /k6EOrckWFuz7I4z4wiRwz8zsj4H.jpg  245000000   

                                              genres  \
0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                            homepage      id    imdb_id  \
0  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   

  original_language                original_title  \
0                en  Star Wars: The Force Awakens   

                                            overview  popularity  \
0  Thirty years after defeating the Galactic Empi...      46.289   

                        poster_path  \
0  /wqnLdwVXoBjKibFRR5U3y0aDUhs.jpg   

                                production_companies  \
0  [{'id': 1, 'logo_path': '/o86DbpburjxrqAzEDhXZ...   

                                production_countries release_date     revenue  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   2015-12-15  2068223624   

   runtime                                   spoken_languages    status  \
0      136  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                         tagline                         title  video  \
0  Every generation has a story.  Star Wars: The Force Awakens  False   

   vote_average  vote_count  belongs_to_collection_id  \
0           7.4       15717                        10   

  belongs_to_collection_name belongs_to_collection_poster_path  \
0       Star Wars Collection  /iTQHKziZy9pAAY4hHEDCGPaOvFC.jpg   

  belongs_to_collection_backdrop_path  
0    /d8duYyyC9J5T825Hg7grmaabfxQ.jpg

In [22]:
pd.json_normalize(data, record_path = 'genres', meta = 'id', meta_prefix = 'movies_')

id             name movies_id
0   28           Action    140607
1   12        Adventure    140607
2  878  Science Fiction    140607
3   14          Fantasy    140607

## Working with APIs and JSON (Part 2)

5. __API-Request__ (discover module): Load all movies with __release date between 2020-01-01 and 2020-02-29__ into a Pandas DataFrame. <br>
See https://www.themoviedb.org/documentation/api/discover and https://developers.themoviedb.org/3/discover/movie-discover for more information.

'f87c31e443414f56263b3306d1fada82'

##  Importing and Saving the Movies Dataset (Best Practice)

6. __API-Request__ (movie module): Load all available information for the movies with movie id = [__299534, 19995, 140607, 299536, 597, 135397, 420818, 24428, 168259, 99861, 284054, 12445, 181808, 330457, 351286, 109445, 321612, 260513__] into a Pandas DataFrame and __save the dataset in a local json file__.

# +++++++++ See some Hints below +++++++++++++

# ++++++++++++++ Hints +++++++++++++++++++++

__Hints for 1.__ <br>
To load json files you can use 

In [ ]:
with open("filename.json") as f:
    data = json.load(f)

and 

In [ ]:
pd.DataFrame(data), pd.read_json(filename.json), pd.json_normalize(data)

the json files have the following orientation (important when using pd.read_json()):
- blockbusters.json -> record
- blockbusters2.json -> column
- blockbusters3.json -> split 

__Hints for 4., 5., 6.__<br>
Make API GET-requests with the library requests (import requests):

In [ ]:
data = requests.get(url).json()

__Hints for 4. and 6.,__ <br> url structure for movie module:

"https://api.themoviedb.org/3/movie/insert_movie_id?api_key=insert_api_key" (replace "insert_movie_id" with movie id and "insert_api_key" with your personal api-key)

__Hints for 5.__<br>
url structure for discover module:

"https://api.themoviedb.org/3/discover/movie?api_key=insert_api_key&query1&query2..." (replace "insert_api_key" with your personal api-key and add appropriate queries)